### ISSUE TITLE
Fmod Update? please??

### ISSUE THREAD

**Issue by Cyndanera** [panda3d/panda3d#384]
Can you Please update fmod ex to the latest version of fmod?

**Comment by Moguri** [comment]
@rdb do we need to be able to support both FMod and FMod Ex? I don't know what kind of API/compatibility concerns there are.

**Comment by rdb** [comment]
@Moguri The original "FMOD Ex" became the new "FMOD Low-Level API".  It is largely the same API and I have a patch somewhere to be able to compile with the new FMOD API.  However, they have changed the parameterisation of the various DSP effects quite a bit, and Panda's API is quite closely following the parameterisation of the old FMOD Ex filters.  So without having some unit tests I've been feeling a bit nervous to check in these changes for fear of breaking existing applications using FMOD filters.

In the long run, there is no reason to keep supporting FMOD Ex, especially as FMOD offers better licensing terms for indie games.  In the short term, we can probably support compiling with whichever one is present in the thirdparty packages.

**Comment by blucheeze-ops** [comment]
@rdb Curious: Why do we currently only allow configuring DSP effects on a channel group level. Why not allow configuring a DSP effect on a particular sound? The FMOD API allows this.

Follow-up question: Would it be ok if for the Core API update we include the ability to configure filters at a per sound level? Meaning I would expose a similar API to configure filters on a sound object? I think this can be done in a way that only extends the API of AudioSound (no breaking changes introduced). 

Something like this:
Current API — bus level
```
fp = FilterProperties()
fp.add_lowpass(800.0, 1.0)
fp.add_echo(0.0, -6.0, 300.0, 40.0)

# Affects every sound routed through this manager.
base.sfxManagerList[0].configure_filters(fp)

```

Hypothetical API — per sound

```
fp = FilterProperties()
fp.add_lowpass(800.0, 1.0)
fp.add_echo(0.0, -6.0, 300.0, 40.0)

gunshot = loader.loadSfx("gunshot.wav")
music   = loader.loadSfx("theme.ogg")
gunshot.configure_filters(fp)
gunshot.play()
music.play()
```

**Comment by blucheeze-ops** [comment]
# FMOD Core API Transition Progress Update
@AVieraTTR and I are getting very close to opening a PR to get FMOD Core API transition complete in Panda3d. We are pretty stoked!

We are currently testing the updates to ensure they work as expected. We are also testing on all platforms: Windows, MacOS, Linux.

We are also working on adding new demos to the `samples/` folder to showcase the sound capabilities (3d, DSPs, etc) of the game engine! We felt sound focused demos were lacking.

## Issue Found
But we have come across a limitation of the DSP Filters API of AudioManager in the game engine. Perhaps intentional. Perhaps not.

In order to change the active DSPs applied to an AudioManager, we must call `configure_filters(fp)` causing the active DSP chain to be rebuilt. This is fine in most cases, especially if you don't find yourself having to make changes quickly.

However, it becomes problematic for at least some DSPs that I have tested when I want to make quick changes to active DSPs.

Take for example the lowpass DSP. Say I am interested in making a sound become clear over time quickly. The only way to do this is to repeated call configure_filters. This as stated before causes the chain of DSPs to be rebuilt entirely. An unfortunate symptom of that is a popping noise. 

You can listen to the issue from this recording I made with my phone while playing the theme song from Cowboy Bebop on my laptop using a similar python script like the example below:
[Tank_lowpass_sweep.mp3](https://github.com/user-attachments/files/25733757/Tank_lowpass_sweep.mp3)

Example:
```python
...
sfx = base.loader.loadSfx("my_sound.wav")
sfx.setLoop(True)
sfx.play()

mgr = base.sfxManagerList[0]
fp = FilterProperties()
t = 0

# Should produce a "muffled" to "clear" and back effect
while True:
    cutoff = 200 + (math.sin(t) * 0.5 + 0.5) * 7800
    fp.clear()
    fp.add_lowpass(cutoff, 1.0)
    mgr.configure_filters(fp)
    t += 0.05
    time.sleep(0.0167)  # Simulate ~60fps updates
``` 

### Proposed Fix

I believe the intended way to update parameters of active DSPs in FMOD API is by simply calling the corresponding `setParameter` method on existing DSP objects. 

What `configure_filters` does is:
1. remove all DSPs in the channel group (`removeDSP`). We destroy the objects (`dsp->release()`). 
2. Make new DSP objects (`createDSPByType`). 
3. Set their parameters (`setParameter`)
4. Add the DSPs to the channel group (`addDSP`)

I propose a new `update_filters` method in addition to `configure_filters` which does one thing:
1. Sets parameters for an existing set of DSPs in a channel group (`setParameter`)

With this, we need a way to easily make adjustments to active DSPs. I propose these extensions to the FilterProperties API to enable this:
```python
mgr = base.sfxManagerList[0]
fp = FilterProperties()

# Slot indices
LOWPASS_INDEX = 0
ECHO_INDEX    = 1

# Build chain structure once — order here defines the slot indices
fp.add_lowpass(200.0, 1.0)
fp.add_echo(0.8, 0.3, 100.0, 0.5)
mgr.configure_filters(fp)

t = 0
while True:
    cutoff = 200 + (math.sin(t) * 0.5 + 0.5) * 7800
    delay  = 50  + (math.sin(t) * 0.5 + 0.5) * 450

    # Use the corresponding index to update parameters of a specific DSP stored in the FilterProperties object.
    # If the index value does not correspond to the DSP type of the update method, the call will return False.
    fp.update_lowpass(LOWPASS_INDEX, cutoff, 1.0)
    fp.update_echo(ECHO_INDEX, 0.8, 0.3, delay, 0.5)

    # Update the filters. We simply continue to reuse the existing FilterProperties object.
    # If the FilterProperties object chain does not match the active chain in the AudioManager, this call
    #      will return False and not apply the updates.
    mgr.update_filters(fp)

    # Read back what is currently set (Just demonstrating this. Not necessary for this example).
    # If the index value does not correspond to the DSP type of the get config method, 
    #     the call will return an empty list of parameters.
    lp  = fp.get_lowpass_config(LOWPASS_INDEX)
    eco = fp.get_echo_config(ECHO_INDEX)
    print(f"lowpass cutoff: {lp[0]:.0f} Hz  |  echo delay: {eco[2]:.0f} ms")

    t += 0.5
    time.sleep(0.0167)
```

With this fix in place, lowpass can now be updated quickly. Take a listen: 
[Tank_lowpass_sweep_update.mp3](https://github.com/user-attachments/files/25734707/Tank_lowpass_sweep_update.mp3)

I have also noticed this fixes an issue with the new Pan DSP. When we update it quickly using `configure_filters`, it fails to register any audible difference when panning the audio. By using the new `update_filters` method, it works as expected. 

To summarize the solution, here is the extended API for `FilterProperties` and `AudioManager`:

`AudioManager`:
- `update_filters(FilterProperties) -> bool`

`FilterProperties`:
- `update_DSP_TYPE(index, params, ...) -> bool`
- `get_DSP_TYPE_config(index) -> list`
- `get_num_filters( ) -> int`

These updates will avoid introducing breaking changes to existing users of Panda3d. The behavior of `configure_filters` and `add` methods of `FilterProperties` remain unchanged. I would be happy to take feedback and suggestions on this or alternative solutions :).

## New DSP Effects
Current list of DSPs supported in the game engine using FMOD Ex:
```cpp
 enum FilterType {
    FT_lowpass,
    FT_highpass,
    FT_echo,
    FT_flange,
    FT_distort,
    FT_normalize,
    FT_parameq,
    FT_pitchshift,
    FT_chorus,
    FT_sfxreverb,
    FT_compress,
  };
```

The new list of DSPs we plan to support with FMOD Core API transition in the game engine:
```cpp
  enum FilterType {
    FT_lowpass,
    FT_highpass,
    FT_echo,
    FT_flange,
    FT_distort,
    FT_normalize,
    FT_parameq,
    FT_pitchshift,
    FT_chorus,
    FT_sfxreverb,
    FT_compress,
    FT_fader,      // New
    FT_limiter,    // New
    FT_pan,        // New
    FT_tremolo,    // New
    FT_delay,      // New
  };
```

**Comment by rdb** [comment]
I don't mind extending the filters to also work on a per-sound level, if the FMOD API trivially allows that. What happens if filters are applied to both the sound and the manager? Do the filters stack?

As for the ability to update existing filters: the proposed API change looks reasonable. In the long term, a more OOP filter definition system might be better, but that would balloon the scope of this change quite a bit.

Implementing get_DSP_config might be a bit tricky with the binding generator.

The DSP-related changes should probably be in a separate PR so as not to delay the basic FMOD API update changes.

**Comment by blucheeze-ops** [comment]
> I don't mind extending the filters to also work on a per-sound level, if the FMOD API trivially allows that. What happens if filters are applied to both the sound and the manager? Do the filters stack?
> 
> As for the ability to update existing filters: the proposed API change looks reasonable. In the long term, a more OOP filter definition system might be better, but that would balloon the scope of this change quite a bit.
> 
> Implementing get_DSP_config might be a bit tricky with the binding generator.
> 
> The DSP-related changes should probably be in a separate PR so as not to delay the basic FMOD API update changes.

Yes, the filters simply stack if applied to both. FMOD makes this trivial.

I will file 2 separate issues for the DSP changes (Per sound sfx and DSP updates). Will open a PR for the FMOD API update soon.

I have get_DSP_config already working. :)
